stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -84,10 +87,55 @@ import SwiftShims
 //   +---------------------------------------+
 //
 //
+// The Native Kinds Buffer
+// -----------------------
+//
+// There are three different classes that can provide native backing storage: 
+// * `_RawNativeDictionaryBuffer`
+// * `_TypedNativeDictionaryBuffer<K, V>`      (inherits from Raw)
+// * `_NativeDictionaryBuffer<K: Hashable, V>` (inherits from Typed)
+//
+// In a less optimized implementation, the parent classes of NativeBuffer could
+// be eliminated, as they exist only to provide special-case behaviours. 
+// NativeBuffer has everything a full implementation of a Dictionary
+// requires, and is subsequently able to provide a full NSDictionary 
+// implementation. Note that NativeBuffer must have the `K: Hashable` constraint
+// because the NSDictionary implementation can't be provided in a constrained 
+// extension.
+//
+// In normal usage, you can expect the backing storage of a Dictionary to be a 
+// NativeBuffer.
+//
+// TypedBuffer is distinguished from NativeBuffer to allow us to create a
+// `NativeStorage<AnyObject, AnyObject>`. Without the Hashable requirement,
+// such a Storage is restricted to operations which can be performed
+// with only the structure of the Buffer: indexing and iteration. This is used
+// in _SwiftDeferredNSDictionary to construct twin "native" and "bridged"
+// storages. Key-based lookups are performed on the native storage, with the
+// resultant index then used on the bridged storage.
+//
+// The only thing that TypedBuffer adds over RawBuffer is an implementation of
+// deinit, to clean up the AnyObjects it stores. Although it nominally 
+// inherits an NSDictionary implementation from RawBuffer, this implementation 
+// isn't useful and is never used. 
+//
+// RawBuffer exists to allow a type-punned empty singleton Buffer to be created.
+// Any time an empty Dictionary is created, this Buffer is used. If this type
+// didn't exist, then NativeStorage would have to store a Buffer that declared
+// its actual type parameters. Similarly, the empty singleton would have to
+// declare its actual type parameters. If the singleton was, for instance, a
+// `NativeBuffer<(), ()>`, then it would be a violation of Swift's strict 
+// aliasing rules to pass it where a `NativeBuffer<Int, Int>` was expected.
+//
+// It's therefore necessary for several types to store a RawBuffer, rather than
+// a TypedBuffer, to allow for the possibility of the empty singleton. RawBuffer
+// also provides an implementation of an always-empty NSDictionary.
+//
+//
 // Index Invalidation
 // ------------------
 //
-// FIXME? not sure if this guarantee is worth making, as it restricts
+// FIXME: decide if this guarantee is worth making, as it restricts
```

Comment:
Sounds good. As far as links go, much of the literature of RHH'ing is bogged down in weird stuff that no real implementation does. Would it be acceptable to link to Rust's implementation notes?

e.g. https://github.com/rust-lang/rust/blob/ac919fcd9d4a958baf99b2f2ed5c3d38a2ebf9d0/src/libstd/collections/hash/map.rs#L70-L178